File: app/src/main/java/com/paxees/tcc/views/adapters/RecyclerViewAdapter.java
package com.paxees.tcc.views.adapters;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.paxees.tcc.R;
import com.paxees.tcc.network.networkmodels.response.models.Category;

import java.util.ArrayList;

import androidx.recyclerview.widget.RecyclerView;

public class RecyclerViewAdapter extends RecyclerView.Adapter<RecyclerViewAdapter.ViewHolder> {
    private int row_index;
    private ArrayList<Category> mData;
    private LayoutInflater mInflater;
    private ItemClickListener mClickListener;
    private Context context;

    // data is passed into the constructor
    public RecyclerViewAdapter(Context context, ArrayList<Category> data) {
        this.mInflater = LayoutInflater.from(context);
        this.mData = data;
        this.context = context;
    }

    // inflates the row layout from xml when needed
    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = mInflater.inflate(R.layout.item_views, parent, false);
        return new ViewHolder(view);
    }

    // binds the data to the TextView in each row
    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        Category animal = mData.get(position);
        Glide.with(context).load(animal.getCategoryIcon()).into(holder.img);
        holder.txt.setText(animal.getCategoryName());
        holder.itemView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                row_index = position;
                notifyDataSetChanged();
            }
        });
        if (row_index == position) {
            holder.img.setBackground(this.context.getResources().getDrawable(R.drawable.button_border));
        } else {
            holder.img.setBackground(this.context.getResources().getDrawable(R.drawable.bg_border));
        }

    }

    // total number of rows
    @Override
    public int getItemCount() {
        return mData.size();
    }


    // stores and recycles views as they are scrolled off screen
    public class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        ImageView img;
        TextView txt;

        ViewHolder(View itemView) {
            super(itemView);
            img = itemView.findViewById(R.id.imgid);
            txt = itemView.findViewById(R.id.tvId);
            itemView.setOnClickListener(this);
        }

        @Override
        public void onClick(View view) {
            if (mClickListener != null) mClickListener.onItemClick(view, getAdapterPosition());
        }
    }

    // convenience method for getting data at click position
    Category getItem(int id) {
        return mData.get(id);
    }

    // allows clicks events to be caught
    public void setClickListener(ItemClickListener itemClickListener) {
        this.mClickListener = itemClickListener;
    }

    // parent activity will implement this method to respond to click events
    public interface ItemClickListener {
        void onItemClick(View view, int position);
    }
}

File: app/src/main/java/com/paxees/tcc/views/adapters/RecyclerViewAdapter.kt
package com.paxees.tcc.views.adapters

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.paxees.tcc.R
import com.paxees.tcc.models.Plants
import java.util.*

class RecyclerViewAdapter(context: Context, data: ArrayList<Plants>, pos: Int) : RecyclerView.Adapter<RecyclerViewAdapter.ViewHolder>() {
    private var row_index = 0
    private val mData: ArrayList<Plants>
    private var view: View? = null
    private val mInflater: LayoutInflater
    private var mClickListener: ItemClickListener? = null
    private val context: Context
    private var pos = 0

    // inflates the row layout from xml when needed
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        view = if (pos == 1) {
            mInflater.inflate(R.layout.item_views_findus, parent, false)
        }else if (pos == 2) {
            mInflater.inflate(R.layout.item_views_diagnose, parent, false)
        } else {
            mInflater.inflate(R.layout.item_views, parent, false)
        }
        return ViewHolder(view!!)
    }

    // binds the data to the TextView in each row
    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val animal = mData[position]
        holder.plantsId.text = animal.plantValue
        holder.itemView.setOnClickListener {
            this.mClickListener?.onItemClick(view, position)
            row_index = position
            notifyDataSetChanged()
        }
        if (pos == 0 || pos == 1) {
            if (row_index == position) {
                holder.plantsId.background = context.resources.getDrawable(R.drawable.bg_border_square_green)
            } else {
                holder.plantsId.background = context.resources.getDrawable(R.drawable.bg_border_square_black)
            }
        } else if (pos == 2) {
            if (row_index == position) {
                holder.plantsId.background = context.resources.getDrawable(R.drawable.bg_border_square_green_diagnose)
            } else {
                holder.plantsId.background = context.resources.getDrawable(R.drawable.bg_border_square_diagnose)
            }
        }
    }

    // total number of rows
    override fun getItemCount(): Int {
        return mData.size
    }

    // stores and recycles views as they are scrolled off screen
    inner class ViewHolder internal constructor(itemView: View) : RecyclerView.ViewHolder(itemView), View.OnClickListener {
        var plantsId: TextView
        override fun onClick(view: View) {
            if (mClickListener != null) mClickListener!!.onItemClick(view, adapterPosition)
        }

        init {
            plantsId = itemView.findViewById(R.id.plantsId)
            itemView.setOnClickListener(this)
        }
    }

    // convenience method for getting data at click position
    fun getItem(id: Int): Plants {
        return mData[id]
    }

    // allows clicks events to be caught
    fun setClickListener(itemClickListener: ItemClickListener?) {
        mClickListener = itemClickListener
    }

    // parent activity will implement this method to respond to click events
    interface ItemClickListener {
        fun onItemClick(view: View?, position: Int)
    }

    // data is passed into the constructor
    init {
        mInflater = LayoutInflater.from(context)
        mData = data
        this.context = context
        this.pos = pos
    }
}